### ISSUE TITLE
Batch Multinomial Function for TensorFlow

### ISSUE THREAD

**Issue by LeavesBreathe** [tensorflow/tensorflow#456]
Hey TF,

Is there some sort of multinomial function that can handle batches? When generating content, it is incredibly useful to introduce variance by 'rolling a dice'. I searched extensively in the docs and could not find one.

[Numpy has one](http://docs.scipy.org/doc/numpy-1.10.1/reference/generated/numpy.random.multinomial.html), but it can not do batches. Tensorflow could capitalize on the fact that it uses a GPU. Thanks alot!

**Comment by girving** [comment]
We don't have this at the moment, but it would be good to add.  Since the right way to do multinomial sampling is [ziggurat](https://en.wikipedia.org/wiki/Ziggurat_algorithm) which uses rejection sampling, the infrastructure to do threaded and GPU-ified random distributions needs to be tweaked to allow multiple numbers of samples.

**Comment by girving** [comment]
For now, depending on how multi you are, quantizing `normal_normal` may be a suffiicent workaround.

**Comment by LeavesBreathe** [comment]
@girving I apologize, but when you say `normal_normal` what exactly are you referring to? I couldn't find a function named that in tensorflow. 

To clarify, I want to take a tensor [batch_size x logits_probabilties] and perform a multinomial function on each one to see which logit is selected. This would result in [batch_size x one_logit_selected].

Thanks again!

**Comment by girving** [comment]
:)  Sorry, I meant `random_normal`.  And my `random_normal` suggestion is not what you want in any case, since you're talking about a categorical distribution, not a multinormial distribution.  I'm in the middle of something, but I may (or may not) have a more useful comment in a couple hours.

**Comment by LeavesBreathe** [comment]
That would be great! Using np.mulitnomial has been pretty challenging due to the fact that the batch size isn't evaluate until you run the session.

**Comment by zheng-xq** [comment]
From an internal discussion, a workaround was suggested using Gumbel argmax trick as follows. We are still discussing whether we should include it in the python op library, or having a more efficient native implementation. 

---------------------------------------------------------

Suppose we want to sample a bunch (say, N) of K-ary categorical distributions. Suppose further that none of the probabilities are exactly zero (to simulate a zero probability for one state just use a really really really small one).

So we have an N by K matrix of of log probabilities X. So in other words we imagine our categorical distributions being expressed in softmax form as exp(X). So really we want to sample from exp(x_i) / sum_j exp(x_j).

Generate a matrix of uniforms on (0,1) of the same shape as X. Call this matrix U.

np.argmax(X - np.log(-np.log(U)), axis=1) gives us our samples. 

https://hips.seas.harvard.edu/blog/2013/04/06/the-gumbel-max-trick-for-discrete-distributions/

**Comment by LeavesBreathe** [comment]
@zheng-xq Thank you for the work-around. If this can be implemented in tensorflow `tf.multinomial` it would be much, much appreciated.

The main problem as I said above is that you have to run `.eval()` to input your tensor into a numpy function. Much of our code is not designed to be evaluated at that point. So having a solution that specifically handles tensorflow's `tensor` would be amazing! Thank you for the help regardless.

**Comment by zheng-xq** [comment]
To clarify, although the quoted workaround uses numpy function as an example. They could be implemented in TensorFlow using their TF counterpart: tf.argmax and tf.log, without invoking eval(). 

And yes, we will likely to make something similar to tf.multinomial at some point. We are still discussing how we should do it.

**Comment by LeavesBreathe** [comment]
> tf.argmax and tf.log, without invoking eval().

Oh duh...I apologize as I should've thought of that. Big Thanks.

**Comment by LeavesBreathe** [comment]
In case anyone is looking at this, here is the sampling function I use. Thanks @zheng-xq  again for all your help.

```python
def batch_sample_with_temperature(a, temperature=1.0):
	'''this function is like sample_with_temperature except it can handle batch input a of [batch_size x logits] 
		this function takes logits input, and produces a specific number from the array. This is all done on the gpu
		because this function uses tensorflow
		As you increase the temperature, you will get more diversified output but with more errors (usually gramatical if you're 
			doing text)
	args: 
		Logits -- this must be a 2d array [batch_size x logits]
		Temperature -- how much variance you want in output
	returns:
		Selected number from distribution
	'''

	'''
	Equation can be found here: https://en.wikipedia.org/wiki/Softmax_function (under reinforcement learning)
        Karpathy did it here as well: https://github.com/karpathy/char-rnn/blob/4297a9bf69726823d944ad971555e91204f12ca8/sample.lua'''
	'''a is [batch_size x logits]'''
	with tf.op_scope(a+temperature, "batch_sample_with_temperature"):
		
		exponent_raised = tf.exp(tf.div(a, temperature)) #start by reduction of temperature, and get rid of negative numbers with exponent
		
		matrix_X = tf.div(exponent_raised, tf.reduce_sum(exponent_raised, reduction_indices = 1)) #this will yield probabilities!

		matrix_U = tf.random_uniform([batch_size, tf.shape(a)[1]], minval = 0, maxval = 1)

		final_number = tf.argmax(tf.sub(matrix_X - matrix_U), dimension = 1) #you want dimension = 1 because you are argmaxing across rows.

	return final_number
```

**Comment by concretevitamin** [comment]
Closing the loop on this thread: we now have a `tf.multinomial()` Op.

**Comment by zxteloiv** [comment]
Thanks for sharing the useful information.
I'm trying to implement the same thing in chainer because it doesn't provide functions to sample a batch of multinomial distributions.

However, in my experiments, the code provided in https://github.com/tensorflow/tensorflow/issues/456#issuecomment-165185025 is wrong, which doesn't give the correct samples. The main problem is that uniform noise is **directly substracted** from the softmax probabilities then take the argmax, differing to a *usual way* that finding index for a uniform noise from ranges defined by the multinomial CDF. By the way, the method provided by https://github.com/tensorflow/tensorflow/issues/456#issuecomment-164038802 is exactly what we want.

Since testing random number generators is not easy, I use the simplest way to sample the same distribution for a million times and count their occurrence. Here's the testing snippet.

```python
import numpy as np

def multinomial_issue_impl(logits):
    exponent = np.exp(logits)
    matrix_X = exponent / np.sum(exponent)
    matrix_U = np.random.rand(*matrix_X.shape)
    final_num = np.argmax(matrix_X - matrix_U)
    return final_num


def multinomial_gumbel_impl(logits):
    exponent = np.exp(logits)
    log_prob = np.log(exponent / np.sum(exponent))
    U = np.random.rand(*log_prob.shape)
    num = np.argmax(log_prob - np.log(-np.log(U)))
    return num


def test(func):
    logits = [1., 1., 1., 1., 1., 5.]  # 0 ~ 5
    K = 1000000

    count = [0] * 6
    for i in xrange(K):
        num = func(logits)
        count[num] += 1

    print count


if __name__ == "__main__":
    print "test:"
    test(multinomial_issue_impl)
    print "gumbel:"
    test(multinomial_gumbel_impl)
```

which yields

```text
test:
[4525, 4415, 4449, 4395, 4451, 977765]
gumbel:
[16674, 16570, 16929, 16902, 16878, 916047]
```

Given the logits `[1., 1., 1., 1., 1., 5.]`, the probabilities should be `[0.01677904,  0.01677904,  0.01677904,  0.01677904,  0.01677904, 0.91610478]`. Thus the result tells that gumbel trick is correct.
